src/shogun/mathematics/graph/Tensor.h

```diff
@@ -24,96 +25,104 @@ namespace shogun
 {
 	namespace graph
 	{
+		/* Creates a copy of the data or a owning data manager on
+		 * the device
+		 */
+		inline std::shared_ptr<ShogunStorage> device_put(
+		    void* ptr, const Shape& shape,
+		    const std::shared_ptr<NumberType>& type, const bool copy)
+		{
+			return std::make_shared<ShogunStorage>(
+			    ptr, shape, type, copy, ShogunStorage::Copy{});
+		}
+
+		/* Creates a view to a pointer. This that the devide
+		 * memory manager will not call the destructor.
+		 */
+		inline std::shared_ptr<ShogunStorage> device_view(
+		    void* ptr, const Shape& shape,
+		    const std::shared_ptr<NumberType>& type)
+		{
+			return ShogunStorage::create_view(ptr, shape, type);
+		}
+
 		class Tensor
 		{
 		public:
+			friend std::shared_ptr<Tensor>
+			from_device(const std::shared_ptr<ShogunStorage>& storage);
+
 			template <typename T>
-			Tensor(const T& scalar)
-			    : m_free(true), m_data(new T(scalar)), m_shape({}),
-			      m_type(from<T>())
+			Tensor(const T& scalar) : m_shape({}), m_type(from<T>())
 			{
+				m_data = device_put(new T(scalar), m_shape, m_type, true);
 			}
 
 			template <typename T>
 			Tensor(const SGVector<T>& vec)
-			    : m_free(false), m_data(vec.vector), m_shape(Shape{vec.size()}),
-			      m_type(from<T>())
+			    : m_shape(Shape{vec.size()}), m_type(from<T>())
 			{
+				m_data = device_put(vec.vector, m_shape, m_type, true);
 			}
 
 			template <typename T>
 			Tensor(const SGMatrix<T>& matrix)
-			    : m_free(false), m_data(matrix.matrix),
-			      m_shape(Shape{matrix.num_rows, matrix.num_cols}),
+			    : m_shape(Shape{matrix.num_rows, matrix.num_cols}),
 			      m_type(from<T>())
 			{
+				m_data = device_put(matrix.matrix, m_shape, m_type, true);
 			}
 
-			~Tensor()
+			template <typename T>
+			Tensor(SGVector<T>&& vec)
```

Comment:
@vigsterkr and also this one which move memory to the storage, rather than allocate+copy